Ensure that air units with less movement left is killed first

trevan opened the pull request:
<!-- If multiple commits please summarize the change above. -->

## Testing
<!-- Describe any manual testing performed below. -->
Stepped through a few battles with different air units with different movements left.  Ensured that the units with less movement left was selected.

## Screens Shots
<!-- If there are UI updates, include screenshots below -->

## Additional Notes to Reviewer
<!-- Add any additional details that would be helpful to reviewers -->

## Release Note

<!--
Include a release note if there is a bug fix or a visible change for players.
For format & syntax help, see:
https://github.com/triplea-game/triplea/blob/master/docs/development/reference/pr-release-notes.md
-->

<!--RELEASE_NOTE-->CHANGE|Air units with less movement is killed before units with more movement.<!--END_RELEASE_NOTE-->

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtySelector.java:
```diff
@@ -254,6 +260,38 @@ private static void killAmphibiousFirst(final List<Unit> killed, final Collectio
     }
   }
 
+  /** Kill off the air units with the lowest movement */
```
1.a)  What are your thoughts to make this method functional? To explain, instead of mutating parameters, the units that should be killed would be returned.

1.b) Otherwise, if not made function, would you mind documenting the side effects of this method, how are the parameters mutated exactly?

2) Documentation, why do we only kill off air units with the lowest movement? A mention that air units are the only units that can move after a battle is the key and that is why we don't do this for land units.

3) Consider extracting the early termination condition to the caller, eg:
```
if(killed.stream().filter(Matches.unitIsAir()).anyMatch()) {
          killLowestMovementAirUnit(...);
}
```

I do think it's a best practice to have as much of the control flow at the top layers and have them call into functional pieces. This means a person can follow the core logic without jumping into additional methods and the additional methods are generally side-effect free and single purpose to accomplishing some task.  <https://github.com/kbilsted/Functional-core-imperative-shell/blob/master/README.md>

4. What were your considerations for (or rather not) unit testing this?  Not worth it, too much of a PITA to create unit objects, or just short on time maybe?

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtySelector.java:
```diff
@@ -254,6 +260,38 @@ private static void killAmphibiousFirst(final List<Unit> killed, final Collectio
+    final Map<UnitOwner, List<Unit>> killedGroupedByOwnerAndType =
+        airUnitsKilled.stream().collect(Collectors.groupingBy(UnitOwner::new, Collectors.toList()));
+
+    for (final Map.Entry<UnitOwner, List<Unit>> killedEntry :
```
Minor, this for loop could perhaps benefit with a short summary comment to help guide the reader. Ideally a short outline of the algorithm or a few words on what we are doing. For example, the 'pop' at the end is not immediately obvious why it is needed, anything to help a reader understand this faster and not reverse engineer the intent is a win.

DanVanAtta reviewed (commented):
Please have a look at the review comments when you have a chance. Otherwise, this is quite the significant fix! This will save players pain and prevent edits, good stuff!

DanVanAtta commented:
@trevan after grouping sub-projects into folders, some merge conflicts came up. Please be aware that 'game-core' is now in 'game-app/game-core'.

trevan commented:
I've rebased against the new master and force pushed.  But now I can't start triplea.  I get the following error:

java.lang.IllegalArgumentException: input == null!
	at java.desktop/javax.imageio.ImageIO.read(ImageIO.java:1400)
	at org.triplea.swing.JFrameBuilder.getGameIcon(JFrameBuilder.java:105)
	at org.triplea.swing.JFrameBuilder.build(JFrameBuilder.java:76)
	at games.strategy.engine.framework.ui.MainFrame.<init>(MainFrame.java:31)
	at games.strategy.engine.framework.ui.MainFrame.buildMainFrame(MainFrame.java:52)
	at games.strategy.engine.framework.GameRunner.lambda$start$0(GameRunner.java:63)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

@DanVanAtta , I'm assuming that I need to run some gradle command?

trevan commented:
I've reworked the code.  Now the code is in CasualtyDetails and is functional.  There are also tests around it.  And I added the code to handle marines in non-partial retreat situations (from https://forums.triplea-game.org/topic/2677/allow-user-to-specifically-choose-amphibious-offloaded-units-in-battle-chooser/8?_=1614527681453) since it is in the same code.

Cernelius commented on game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtySelector.java:
```diff
@@ -150,13 +150,16 @@ public static CasualtyDetails selectCasualties(
 
-    // if partial retreat isn't possible, then ensure units with marine bonus are removed first or
-    // last depending on whether the bonus is negative or positive
+    // if partial retreat isn't possible and it was an amphibious assault, then ensure units with
```
You should specify here we are talking about partial retreat of land units (the absence of it meaning that land units either can retreat all together or cannot retreat). Retreating air only is a partial retreat too, and I assume whether or not the partial retreat of air only is possible is irrelevant to this matter.

DanVanAtta replied:
It's a good sign you're jumping in with feedback @Cernelius 

The context of this comment is very much the next 10 lines of code. For the more generic documentation, check for things that are in block comments `/** */` and the content in `docs`, comments starting with `//` are going to be about the nearby code.

I'll leave it to @trevan to consider incorporating this feedback further.

With that said, the commentary does help and the next block of code is quite heavy. I wonder if some explanation variables can help and make the commentary unnecessary (commentary is a source of maintenance, so we want it to be terse and concise).

What do you think of something like this @trevan ?
```
boolean amphibRetreatNotPossible = !Properties.getPartialAmphibiousRetreat(data.getProperties();
if(amphibRetreatNotPossible) {
  boolean marineCasualtiesSelected = casualtySelection.getKilled().stream()
                .anyMatch(
                    unit -> unit.getUnitAttachment().getIsMarine() != 0 && unit.getWasAmphibious());
   boolean marinesDamaged = casualtySelection.getDamaged().stream()
                .anyMatch(
                    unit ->
                        unit.getUnitAttachment().getIsMarine() != 0 && unit.getWasAmphibious()));
  if(marineCasualtiesSelected || marinesDamaged) {
     casualtySelection =
          casualtySelection.ensureUnitsWithPositiveMarineBonusAreTakenLast(sortedTargetsToPickFrom);
   }
}
```

It could also be worth considering to rename the various methods/lines to refer to amphibious(assaulting) units in case we have units that are amphibious but not marines (say something like amphibious vehicles).

trevan replied:
I removed the comments and added explanation variables.

By the way, a "marine" could easily be a tank, artillery, etc.  It just has to have the marine bonus set and it would need to have been offloaded by a sea transport before the battle.  So this code isn't specific to a unit called "marine".  It is solely looking at the attributes that are set in the unit definition (isMarine) and the status of being offloaded from a sea transport during the same combat phase (wasAmphibious).

Cernelius commented on game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtySelector.java:
```diff
@@ -150,13 +150,16 @@ public static CasualtyDetails selectCasualties(
-    // if partial retreat isn't possible, then ensure units with marine bonus are removed first or
-    // last depending on whether the bonus is negative or positive
+    // if partial retreat isn't possible and it was an amphibious assault, then ensure units with
+    // marine bonus are removed first or last depending on whether the bonus is negative or positive
```
It would be proper to add a "respectively" here. Without it, it might be unclear that "negative" refers to "first" and "positive" refers to "last". Also, a mention of "amongst the same-named ones", referring to "units" may be added so to avoid that someone might think that, for example, the program will automatically take out infantry before offloaded marines, whereas here we are referring only to taking out non-offloaded marines before offloaded ones.

DanVanAtta commented:
> I've rebased against the new master and force pushed. But now I can't start triplea.

@trevan are you using the IDEA launcher, or CLI?

trevan commented:
> > I've rebased against the new master and force pushed. But now I can't start triplea.
> 
> @trevan are you using the IDEA launcher, or CLI?

@DanVanAtta IDEA launcher

DanVanAtta commented:
downloadAssets is the likely suspect, double check the config looks good in the launcher:

![Screenshot from 2021-02-28 17-10-23](https://user-images.githubusercontent.com/12397753/109440666-e61d0580-79e7-11eb-880d-4b132b991417.png)


That is checked in AFAIK. A `./gradlew clean` might be in order.

trevan commented:
@DanVanAtta , the downloadAssets runs without issues and it looks fine.  I ran the clean command and then tried again and it still has the same problem.

DanVanAtta commented:
Hmm..

Is there any difference, perhaps module or working directory in your run config?

![Screenshot from 2021-02-28 20-51-13](https://user-images.githubusercontent.com/12397753/109453385-c6490a00-7a06-11eb-9f0c-8c2e4728cd31.png)

Is there a physical assets folder created?

```
dan@dan-xps:~/work/triplea$ find . -name "assets"
./game-app/game-headed/assets
```

trevan commented:
✗ find . -name "assets"
./game-app/game-headed/assets

trevan commented:
How do I run the headed from the command line?

DanVanAtta commented:
```
./gradlew :game-app:game-headed:run
```

DanVanAtta commented:
Nice work, thank you @trevan

trevan commented:
@DanVanAtta , running it from the command line gives me the same error.  I tried running clean before hand and then running it from the command line as well.  Still get the error.

trevan commented:
I also tried a git clean beforehand and still have the error.

DanVanAtta commented:
Do you have ta_icon.png any of these locations?

```
dan@dan-xps:~/work/triplea$ find . -name "ta_icon.png"
./lib/swing-lib/out/production/resources/org/triplea/swing/ta_icon.png
./lib/swing-lib/src/main/resources/org/triplea/swing/ta_icon.png
./lib/java-extras/out/production/resources/org/triplea/swing/ta_icon.png
./lib/java-extras/src/main/resources/org/triplea/swing/ta_icon.png
```

trevan commented:
Not in any of those locations:

✗ find . -name "ta_icon.png"
./game-app/game-headed/assets/icons/ta_icon.png

DanVanAtta commented:
I think the 'right' location is swing-lib. IIRC we have png files in gitignore, it would explain why it is not checked in and available for you.

Side-note, the icon is something that ought to be bundled with assets, so it ought to be checked in with the assets repo and removed from the game repo.

trevan commented:
The resource directory didn't even exist in swing-lib.  After creating it and copying the ta_icon.png there, the app finally loaded.
